tpm/debug/pcapng.go

```diff
@@ -0,0 +1,188 @@
+type FlushFunc func() error
+
+// NewPcapngTap creates a new TPM tap that writes the
+// the outgoing and incoming TPM communication to the
+// provided [io.Writer] in PcapNG format.
+func NewPcapngTap(w io.Writer) (Tap, FlushFunc, error) {
+	// record current defaults, so that they can be restored after
+	// creating a PcapNG writer.
+	currentSectionInfo := pcapgo.DefaultNgWriterOptions.SectionInfo
+	defer func() {
+		pcapgo.DefaultNgWriterOptions.SectionInfo = currentSectionInfo
+	}()
+	currentInterface := pcapgo.DefaultNgInterface
+	defer func() {
+		pcapgo.DefaultNgInterface = currentInterface
+	}()
+
+	// override some default properties of the PcapNG writer
+	pcapgo.DefaultNgWriterOptions.SectionInfo = ngSectionInfo
+	pcapgo.DefaultNgInterface = ngInterface
+
+	// create a new PcapNG writer
+	pw, err := pcapgo.NewNgWriter(w, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed creating PcapNG writer: %w", err)
+	}
+	finish := func() error {
+		if err := pw.Flush(); err != nil {
+			return fmt.Errorf("failed to flush PcapNG writer: %w", err)
+		}
+		return nil
+	}
+
+	// the single PcapNG writer is used for both TPM commands and responses.
+	// It's wrapped, so that the direction of the data is know at the time of
+	// writing a new packet to the PcapNG writer.
+	return &pcapTap{
+		in: &pcapWriter{
+			in:     true,
+			writer: pw,
+		},
+		out: &pcapWriter{
+			writer: pw,
+		},
+	}, finish, nil
+}
+
+type pcapTap struct {
+	in  *pcapWriter
+	out *pcapWriter
+}
+
+func (t *pcapTap) Rx() io.Writer {
+	return t.in
+}
+
+func (t *pcapTap) Tx() io.Writer {
+	return t.out
+}
+
+type pcapWriter struct {
+	in     bool
+	writer *pcapgo.NgWriter
+}
+
+var (
+	outSeq uint32
+	inSeq  uint32
+	mu     sync.Mutex
+)
+
+// Write implements [io.Writer] and writes the provided data
+// to the underlying [io.Writer] in PcapNG  format.
+func (w *pcapWriter) Write(data []byte) (int, error) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	err := write(w.writer, data, w.in, inSeq, outSeq)
+
+	if w.in {
+		inSeq += uint32(len(data))
+	} else {
+		outSeq += uint32(len(data))
+	}
+
+	return len(data), err
+}
+
+var ethernetLayer = &layers.Ethernet{
+	SrcMAC:       net.HardwareAddr{0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+	DstMAC:       net.HardwareAddr{0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+	EthernetType: layers.EthernetTypeIPv4,
+}
+
+var ipLayer = &layers.IPv4{
+	Version:  4,
+	TTL:      64,
+	Flags:    layers.IPv4DontFragment,
+	Protocol: layers.IPProtocolTCP,
+	SrcIP:    net.IP{127, 0, 0, 1},
+	DstIP:    net.IP{127, 0, 0, 1},
+}
+
+var serializeOptions = gopacket.SerializeOptions{
+	FixLengths:       true,
+	ComputeChecksums: true,
+}
+
+var decodeOptions = gopacket.DecodeOptions{
+	Lazy:   true,
+	NoCopy: true,
+}
+
+func write(w *pcapgo.NgWriter, data []byte, in bool, inSeq, outSeq uint32) error {
+	var tcpLayer = &layers.TCP{Window: 16}
+	if in {
+		tcpLayer.SrcPort = layers.TCPPort(2321)
+		tcpLayer.DstPort = layers.TCPPort(50001)
```

Comment:
Didn't look too much into this PR, but we probably want to move these ports customizable.

Reply:
We could, but that wouldn't change much functionally. The PCAP is used just as a container in this case; we're not actually capturing traffic from/to those ports. The bytes read from and written to the TPM device are "massaged into network packets", so to say.